src/replay.rs

```diff
@@ -0,0 +1,567 @@
+//! A utility for recording over time
+
+use crate::agent_setup::PacmanAgentSetup;
+use crate::game_state::PacmanState;
+use crate::standard_grids::StandardGrid;
+use anyhow::{anyhow, Error};
+use rapier2d::na::Isometry2;
+use serde::{Deserialize, Serialize};
+use std::time::{Duration, SystemTime};
+
+/// The types of data that might be stored in a [`ReplayFrame`]
+#[derive(Clone, Serialize, Deserialize)]
+enum ReplayFrameData {
+    /// Pacbot's real physical location, as determined by the [`PacbotSimulation`]
+    PacbotLocation(Isometry2<f32>),
+    /// Information that changes frequently in Pacman, like ghost locations and pellets
+    PacmanGameState(Box<PacmanState>),
+}
+
+/// The metadata included in one frame of a [`Replay`]
+#[derive(Clone, Serialize, Deserialize)]
+struct ReplayFrame {
+    /// The data in the frame
+    pub data: ReplayFrameData,
+    /// When the data was created
+    pub timestamp: SystemTime,
```

Comment:
It might be good to use `std::time:Instant` instead of `SystemTime` to get a monotonic clock.